### ISSUE TITLE
Multiple questions for the FAQ, or possibly issues

### ISSUE THREAD

**Issue by andreas-kupries** [tclssg/tclssg#30]
For the background I am currently trying to convert http://www.oche.de/~akupries to TclSSG to eval eval it.

Questions of note:

- How can I specify sidebar links on non-blog pages?
- How can I specify a logo (image) to be shown to the left of the navigation bar ? (I suspect that this might require editing of the template files ?! If yes, which, where, how?)
- I would like to have the copyright year set automatically, and also have it contain a link to a disclaimer page. The text does not seem to allow markdown. (Or maybe with enableMacrosINPage ?) ... A basic link can't be relative, because it is always relative to the current page, so I would need multiple disclaimers for each possible directory.
- Is "bootstrap" the one template which exists? Is there a list of other pre-made templates which could be used ?

Depending on the answers these q's and answers might have to go into the FAQ, or become enhancement requests.

**Comment by andreas-kupries** [comment]
Oh, things I forgot ...

- The blog/tag.md is the template for generated tag pages. Does this template have access to some variable (or command) which tells it the text of the tag the page is currently generated for ? I.e. if I have tags Tcl, SSG, etc. the tag.md will be used for each tag X in turn, and I want to insert the tag X into the page. I.e. Instead of a plain "This is a tag page" I want to write "This is the tag for "X"", where X is the current tag.

**Comment by andreas-kupries** [comment]
You had an example demonstrating "adaptable images" (IIRC). It worked by essentially sourcing a file ... "img-procs.tcl" in each page which needed it. Is there a place to get such a file loaded globally so that it automatically applies to all pages?

**Comment by andreas-kupries** [comment]
It might be useful to have a document which spells out the steps from input the page/blog .md file to .html, with all variables, or templates involved at each step documented as well. And then extended to cover how the blog and tag pages are generated from their .md templates too.

Basically for the semi-advanced use case I currently run into with logo, sidebar, disclaimer, ...

**Comment by dbohdan** [comment]
I'll go over this point by point but first let me point out that Tclssg is at the end of a big pre-1.0 overhaul right now, so the documentation is out of sync with the code. If you intend to use it in production you may want to wait until 1.0.0b is [released](https://github.com/dbohdan/tclssg/releases).

>How can I specify sidebar links on non-blog pages?

You can force-show the sidebar note by setting the variable `showSidebarNote` to 1 and putting the links in the note for the page. This is deliberately undocumented, however, and is not guaranteed to stay in 1.0 because having both `showSidebarNote` and `hideSidebarNote` is confusing. Now that separate sets of page variable defaults for plain old pages and blog posts are configurable in the config ( `pageVariables` and `blogPostVariables`) a more elegant solution might be to use them instead. I'll take care of this in the next commit and document the right approach then. [Issue](https://github.com/dbohdan/tclssg/issues/33).

As an alternative you can put links in the navbar. Perhaps the name "sidebar links" is somewhat misleading because in the documentation and the code it means only the *automatically generated* links section on blog posts. User links go in the sidebar *note*.

>How can I specify a logo (image) to be shown to the left of the navigation bar ? (I suspect that this might require editing of the template files ?! If yes, which, where, how?)

There is no way to do it via the config now. You could edit `bootstrap.thtml` to include the logo in the navbar manually (see http://stackoverflow.com/questions/18474564/bootstrap-3-navbar-with-logo). This seems an overlooked essential feature, so I've added a [separate issue](https://github.com/dbohdan/tclssg/issues/31) for it.

>I would like to have the copyright year set automatically, and also have it contain a link to a disclaimer page. The text does not seem to allow markdown. (Or maybe with enableMacrosINPage ?) ... A basic link can't be relative, because it is always relative to the current page, so I would need multiple disclaimers for each possible directory.

Great question. For now I can only suggest to use a fixed copyright year and absolute links. I'll get back to you on this when I figure out a good way to do this. [Issue](https://github.com/dbohdan/tclssg/issues/32).

>Is "bootstrap" the one template which exists? Is there a list of other pre-made templates which could be used ?

There are two sets of document and article templates: Bootstrap (`skeleton/templates/*.thtml`) and RSS (`skeleton/templates/*.txml`). There used to be two more for HTML but they are not compatible with the current version of Tclssg (and Bootstrap is prettier). If you only want a different look, rather different markup, consider using a [Bootstrap](http://bootstrapzero.com/) [theme](http://startbootstrap.com/).

>The blog/tag.md is the template for generated tag pages. Does this template have access to some variable (or command) which tells it the text of the tag the page is currently generated for ? I.e. if I have tags Tcl, SSG, etc. the tag.md will be used for each tag X in turn, and I want to insert the tag X into the page. I.e. Instead of a plain "This is a tag page" I want to write "This is the tag for "X"", where X is the current tag.

No. There is `[get-current-page-variable tagPageTag {}]` but because of how the pages are processed (macro expansion -> article template -> document template) it can't be used in a page macro. For how to use it in the document template look at proc `format-document-title` in `skeleton/templates/document.tcl`.

>It might be useful to have a document which spells out the steps from input the page/blog .md file to .html, with all variables, or templates involved at each step documented as well. And then extended to cover how the blog and tag pages are generated from their .md templates too.

This is an excellent idea. I will write it up either in the README or on the wiki.

I think these are some very good and useful questions. I will put the answers in the FAQ/README but only once I know they won't change much.

**Comment by dbohdan** [comment]
>You had an example demonstrating "adaptable images" (IIRC). It worked by essentially sourcing a file ... "img-procs.tcl" in each page which needed it. Is there a place to get such a file loaded globally so that it automatically applies to all pages? 

The example is still in the README but `img-procs.tcl` has been renamed `img.tcl`. There is currently no way to source it in all files by default other than copying   `skeleton/common.tcl` to `[your-project]/input/templates/common.tcl` and editing it (`interp-source` is a bit more complicated than plain `source` to allow local project files in `templates/` to override those in `skeleton/templates/`). However, this is something you shouldn't normally do because you'll have to keep it in sync manually when Tclssg is updated. Instead, source  `img.tcl` on the pages you want to use it on as a temporary fix and I'll probably make it included by default later (likely in the next commit). [Issue](https://github.com/dbohdan/tclssg/issues/34).

**Comment by dbohdan** [comment]
P.S.: I suggest you subscribe to the issues I've created in response to your questions so that you get notified when they are fixed or have new comments.

**Comment by andreas-kupries** [comment]
Thanks for taking the time to answer my questions ...
   - I am now subscribed to the 3 issues you made for me.

   - "Tclssg is at the end of a big pre-1.0 overhaul right now".
     That is good to know. I will watch, and better subscribe to the repository for commit notifications.

   - "sidebar links on non-blog pages" -- showSidebarNote, undocumented and changing.
    Ok, will wait with this on what happens with the issue. My plan, which I did not mention before, was to use the sidebar on the normal pages as secondary navigation bar, with per-page content, i.e. context-sensitive. That precludes your proposal of putting them into the nav bar, because that one is global, the same for all pages.

   - "logo"
      Not via config - I suspected. I will have a look at the referenced page and see if I can figure it out.

   - auto copyright and link ... Yesterday evening I tried a macro (<%! ... %>) and can confirm that this is not working either. I will go with fixed for now (and have a year to wait for support in ssg). I will put in an absolute link as well. The inconvenience is when the website gets deployed to multiple locations. All will point to one of them for the disclaimer, i.e. outside of themselves.

   - bootstrap and other templates ... I think my question was about themes for bootstrap, but phrased confusingly. I will have a look at the links you gave me, thanks.

   - tag.md customization ... Ok. That one was pretty much the lowest of the priorities, and I can live without that. (It just annoys me).

   - documenting the processing steps, README or wiki.
    IMHO the wiki would be the better place for this. The README already feels to me to be overly long. IMVHO the parts documenting the various global and page configs could/should also be put on separate wiki pages and then just referenced from the wiki.

    I usually try to separate documentation by audience, i.e. regular user, vs. installer/packager, vs. developer and then write towards that. There is of course some overlap, i.e. a user has to know parts of how things works. On the other hand, only a dev has to know the deep details. An example of that can be found at [Cmdr](http://core.tcl.tk/akupries/cmdr/index). The "Instructions" are mostly installer, packager and dev, and then there are the user docs, reference and intro, with the latter giving the broad overview and pointing to the relevant references from there.

   And yes, writing all that is very much quite a lot of work.

**Comment by dbohdan** [comment]
Regarding `copyright` and the sidebar note: now you can do it. See the commit message and the README changes for https://github.com/dbohdan/tclssg/commit/8e339b3a953d5260ef6b3af04f651bf39288ea5b.

**Comment by dbohdan** [comment]
I'll comment some more tomorrow (EET). For now I want say that I agree with you re: the README being too long.

**Comment by dbohdan** [comment]
I have thought some more about project documentation and looked at your example of Cmdr as well as some other packages. I had considered several ways of restructuring the documentation before but didn't go through with any of them. Having examined other projects' documentation from the perspective of separating documentation by audience I have come to the conclusion that it is a very good idea (and like many really good ideas it is obvious in retrospect). This is likely the optimal solution for Tclssg. Therefore, I want to say "thank you" for insight on this matter.

I'll start gradually to move the documentation from the README to the [wiki](https://github.com/dbohdan/tclssg/wiki). The wiki is open for writing so if you want to contribute definitely feel free to do so. However, until the transfer is complete I'd encourage you to freely create and edit your own pages but be careful with edits to the documentation from the README. For commentary maybe we could sign our comments like on the Tcler's wiki? — I think that's a good process.

It is somewhat unfortunate that the wiki will not be versioned together with the main repo (i.e., like it would be in Fossil) but at least it is not in a proprietary format; its contents is available as a Git repository of simple Markdown files. Hmm, come to think of it, one could use Tclssg to produce a browsable local copy of it.

**Comment by andreas-kupries** [comment]
I started using my quad of (how to get, how to build, how to use, how to modify) about ... 2-3 years ago now I think. It really helps me when writing, especially when it comes to things like "how deep into details should I go here ?". Or "Is that what I am about to write actually relevant to the audience ?"

Re Wiki, sure, if I have write-perms for it I will see where I find the time to make some pages.
Does the wiki allow pages to be renamed or deleted ? In that case I will put "AK" into the title of any page I make and when it becomes official it can either be renamed to remove that, or copied to a new page with the proper title, and mine then deleted.

**Comment by dbohdan** [comment]
>I started using my quad of (how to get, how to build, how to use, how to modify) about ... 2-3 years ago now I think.

This may be the kind of idea for which it is worth having a name ("how to get, how to build, how to use, how to modify" could work) and a "canonical" explanation one can link people to. Have you written about it in more detail somewhere? If not, consider doing so when you have the time and posting it on a blog or, e.g, on the Tcler's wiki. I've read articles on writing documentation that would hint this idea at but not quite explicate it as well as articles on UX that talked about different "user personas" but not in the context of writing documentation. I can also write it up I'd rather only do so once I've had enough experience with the approach to provide meaningful commentary on it.

>Does the wiki allow pages to be renamed or deleted ?  In that case I will put "AK" into the title of any page I make and when it becomes official it can either be renamed to remove that, or copied to a new page with the proper title, and mine then deleted.

Yes to both. Your method should work fine and I think it is a good one.

**Comment by dbohdan** [comment]
I've moved most of the documentation to the wiki organized into "User's guide" (existing documentation), "Template writer's guide" (currently just the note about the processing order) and "Developer's guide" (TBD):  https://github.com/tclssg/tclssg/wiki. Unfortunately, GitHub's web interface can display but not correctly edit different sidebar files in different subdirectories. You have to clone the wiki repository and edit it locally for that. 

Should I close this issue?

**Comment by andreas-kupries** [comment]
Closing - Sure.
Move - Did not realize that despite seeing and accepting your invitation to the org. Thanks for that.
Will have to check how my fork now connects.
